### PR TITLE
Use multistage build to remove zypper stack from final image

### PR DESCRIPTION
**Pull request by dirkmueller**
This can be used for other app containers as well. Unsure if it should be done globally for all application containers.

### CODE REVIEW

**Review by dcermak — CHANGES_REQUESTED** [review]
Sorry, but I don't like this. We have a much better way to create small containers: use kiwi instead of resorting to this multistage hackery.

**Comment by dirkmueller** [comment]
"requests changes" means what needs to be changed? we can't change to kiwi build easily because it changes repositories so qa pipelines etc needs adjustment. 

I don't think kiwi is a better approach. enabling a multistage build is seamless and allows to remove packages that frequently attract vulnerabilities.

**Comment by dcermak** [comment]
> "requests changes" means what needs to be changed? we can't change to kiwi build easily because it changes repositories so qa pipelines etc needs adjustment.

I wouldn't find the change that intrusive, given that this just touches tomcat, that hasn't been setup yet.

> I don't think kiwi is a better approach. enabling a multistage build is seamless and allows to remove packages that frequently attract vulnerabilities.

Kiwi allows us to do the same. One disadvantage of the multistage build is, that the `.packages` file that OBS produces is no longer correct. E.g. [`suse-tomcat-10-15.6.10.x86_64-35.7.packages`](https://build.opensuse.org/build/home:defolos:BCI:Staging:SLE-15-SP6:6-1387/containerfile/x86_64/tomcat-10-image/suse-tomcat-10-15.6.10.x86_64-35.7.packages) contains zypper and all its dependencies. The SBOMs appear to be correct though.

**Comment by dcermak** [comment]
> > I don't think kiwi is a better approach. enabling a multistage build is seamless and allows to remove packages that frequently attract vulnerabilities.
> 
> Kiwi allows us to do the same. One disadvantage of the multistage build is, that the `.packages` file that OBS produces is no longer correct. E.g. [`suse-tomcat-10-15.6.10.x86_64-35.7.packages`](https://build.opensuse.org/build/home:defolos:BCI:Staging:SLE-15-SP6:6-1387/containerfile/x86_64/tomcat-10-image/suse-tomcat-10-15.6.10.x86_64-35.7.packages) contains zypper and all its dependencies. The SBOMs appear to be correct though.

The `report` files suffer from the same flaw as the `packages` files, e.g. [`suse-tomcat-10-15.6.10.x86_64-35.7.report`](https://build.opensuse.org/build/home:defolos:BCI:Staging:SLE-15-SP6:6-1387/containerfile/x86_64/tomcat-10-image/suse-tomcat-10-15.6.10.x86_64-35.7.report) contains the whole zypper stack too. This will mess up the reports that @msmeissn generates.

**Comment by dirkmueller** [comment]
> The `report` files suffer from the same flaw as the `packages` files, e.g. [`suse-tomcat-10-15.6.10.x86_64-35.7.report`](https://build.opensuse.org/build/home:defolos:BCI:Staging:SLE-15-SP6:6-1387/containerfile/x86_64/tomcat-10-image/suse-tomcat-10-15.6.10.x86_64-35.7.report) contains the whole zypper stack too. This will mess up the reports that @msmeissn generates.

Marcus mentioned to me that it would not be an issue for him to switch to SBOMs for reports, which I think would make most sense.

**Review by dcermak — CHANGES_REQUESTED** [review]
I'm not a giant fan of another mass change but I don't know a better way either. My two complaints are:
- the README is now wrong as it suggests to use zypper in two places
- lack of `curl` in the tomcat container

File: src/bci_build/templates.py [hunk 1]
```diff
@@ -39,7 +39,13 @@
 #!BuildRelease: {{ image.build_release }}
 {%- endif %}
 {{ image.dockerfile_from_line }}
+{%- if image.from_target_image %}
+COPY --from=target / /target
```
[the remark this answers is not in the document]
Comment (dirkmueller) [reply src/bci_build/templates.py:43]: yes. see https://opensource.suse.com/bci-docs/guides/use-with-golang/

File: src/bci_build/package/apache_tomcat.py [hunk 2]
```diff
@@ -67,12 +69,14 @@ def _get_sac_supported_until(
...
         ],
+        from_target_image=f"{_build_tag_prefix(os_version)}/bci-micro:{OsContainer.version_to_container_os_version(os_version)}",
         package_list=[
             tomcat_pkg := (
                 "tomcat"
                 if tomcat_major == _TOMCAT_VERSIONS[0]
                 else f"tomcat{tomcat_major}"
-            )
+            ),
+            "sed",
```
[the remark this answers is not in the document]
Comment (dirkmueller) [reply src/bci_build/package/apache_tomcat.py]: merged

**Review by dcermak — APPROVED** [review]

**Pull request merged**